### PR TITLE
Issue #2711 and general UUID improvement

### PR DESCRIPTION
**Pull request by kothiga**
### Changelog category (leave one):
- Improvement
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Changed UUID internal representation to have proper byte continuity. Ensured backwards compatibility with serialization and hashing.

### CODE REVIEW

File: src/Functions/FunctionsHashing.h [hunk 1]
```diff
@@ -1139,45 +1139,44 @@ class FunctionAnyHash : public IFunction
...
+                if constexpr(std::is_same_v<FromType, UUID>)
                 {
-                    char tmp_buffer[sizeof(vec_from[i])];
-                    reverseMemcpy(tmp_buffer, &vec_from[i], sizeof(vec_from[i]));
-                    h = apply(key, reinterpret_cast<const char *>(tmp_buffer), sizeof(vec_from[i]));
+                    auto & uuid = value.toUnderType();
+                    std::swap(uuid.items[0], uuid.items[1]);
                 }
+
+                ToType hash = apply(key, reinterpret_cast<const char *>(&value), sizeof(value));
```
Comment (ltrk2) [inline src/Functions/FunctionsHashing.h:1152]: You could make this const (and even auto).

File: src/Functions/FunctionsHashing.h [hunk 2]
```diff
@@ -1139,45 +1139,44 @@ class FunctionAnyHash : public IFunction
...
-            else
-            {
-                char tmp_buffer[sizeof(value)];
-                reverseMemcpy(tmp_buffer, &value, sizeof(value));
-                h = apply(key, reinterpret_cast<const char *>(tmp_buffer), sizeof(value));
+                auto & uuid = value.toUnderType();
+                std::swap(uuid.items[0], uuid.items[1]);
             }
+
+            ToType hash = apply(key, reinterpret_cast<const char *>(&value), sizeof(value));
```
Comment (ltrk2) [inline src/Functions/FunctionsHashing.h:1171]: You could make this const (and even auto).

**Review by ltrk2 — CHANGES_REQUESTED** [review]

File: tests/queries/0_stateless/02769_uuid_format_serialization.sql [hunk 3]
```diff
@@ -0,0 +1,3 @@
+SELECT toUUID('61f0c404-5cb3-11e7-907b-a6006ad3dba0') FORMAT RowBinary;
+SELECT toUUID('992f6910-42b2-43cd-98bc-c812fbf9b683') FORMAT RowBinary;
+SELECT toUUID('417ddc5d-e556-4d27-95dd-a34d84e46a50') FORMAT RowBinary;
```
[the remark this answers is not in the document]
Comment (kothiga) [reply tests/queries/0_stateless/02769_uuid_format_serialization.sql]: Addressed and improved the functional test.

File: src/IO/WriteHelpers.h [hunk 4]
```diff
@@ -86,6 +86,14 @@ inline void writePODBinary(const T & x, WriteBuffer & buf)
     buf.write(reinterpret_cast<const char *>(&x), sizeof(x)); /// NOLINT
 }
 
+inline void writePODBinary(const UUID & x, WriteBuffer & buf) 
+{
+    auto uuid = x.toUnderType();
+    std::swap(uuid.items[0], uuid.items[1]);
+
+    writePODBinary(uuid, buf);
```
[the remark this answers is not in the document]
Comment (kothiga) [reply src/IO/WriteHelpers.h]: Good suggestion. Applied.

File: src/IO/ReadHelpers.h [hunk 5]
```diff
@@ -113,6 +113,14 @@ inline void readPODBinary(T & x, ReadBuffer & buf)
     buf.readStrict(reinterpret_cast<char *>(&x), sizeof(x)); /// NOLINT
 }
 
+inline void readPODBinary(UUID & x, ReadBuffer & buf)
+{
+    auto & uuid = x.toUnderType();
+    readPODBinary(uuid, buf);
+    
+    std::swap(uuid.items[0], uuid.items[1]);
```
[the remark this answers is not in the document]
Comment (kothiga) [reply src/IO/ReadHelpers.h]: Good suggestion. Applied.

File: src/Functions/FunctionsHashing.h (hunk 1 above)
Comment (kothiga) [reply src/Functions/FunctionsHashing.h]: Applied.

File: src/Functions/FunctionsHashing.h (hunk 2 above)
Comment (kothiga) [reply src/Functions/FunctionsHashing.h]: Applied.

**Review by ltrk2 — APPROVED** [review]

**Comment by kothiga** [comment]
Fixed Issue #2711 and adjusted internal behavior of UUIDs

**Pull request merged**